components/client/src/main/java/com/hotels/styx/client/connectionpool/ConnectionExpirationTracker.java

```diff
@@ -0,0 +1,102 @@
+ */
+package com.hotels.styx.client.connectionpool;
+
+import com.google.common.base.Stopwatch;
+import com.google.common.base.Ticker;
+import com.hotels.styx.api.HttpRequest;
+import com.hotels.styx.api.HttpResponse;
+import com.hotels.styx.api.client.Connection;
+import com.hotels.styx.api.client.Origin;
+import rx.Observable;
+
+import java.util.function.Supplier;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+/**
+ * Provides wrapper for connection, that stores a connection expiration time. Also provides a method for verification of
+ * a time pasted.
+ */
+class ConnectionExpirationTracker implements ConnectionUsageTracker {
+    private final long connectionExpirationSeconds;
+    private final Supplier<Ticker> tickerSupplier;
+
+    ConnectionExpirationTracker(long connectionExpirationSeconds, Supplier<Ticker> tickerSupplier) {
+        this.connectionExpirationSeconds = connectionExpirationSeconds;
+        this.tickerSupplier = tickerSupplier;
+    }
+
+    @Override
+    public ExpirableConnection decorate(Connection connection) {
+        requireNonNull(connection);
+        return new TrackedConnectionAdapter(connection, connectionExpirationSeconds, tickerSupplier);
+    }
+
+    @Override
+    public boolean shouldTerminate(Connection connection) {
+        return connection instanceof ExpirableConnection && ((ExpirableConnection) connection).isExpired();
+    }
+
+    private class TrackedConnectionAdapter implements ExpirableConnection {
+        private final Connection nettyConnection;
+        private final long connectionExpirationSeconds;
+        private final Stopwatch stopwatch;
+
+        TrackedConnectionAdapter(Connection nettyConnection, long connectionExpirationSeconds, Supplier<Ticker> tickerSupplier) {
+            this.connectionExpirationSeconds = connectionExpirationSeconds;
+            this.nettyConnection = nettyConnection;
+            this.stopwatch = Stopwatch.createStarted(tickerSupplier.get());
+        }
+
+        @Override
+        public boolean isConnected() {
+            return nettyConnection.isConnected();
+        }
+
+        @Override
+        public Observable<HttpResponse> write(HttpRequest request) {
+            return nettyConnection.write(request);
+        }
+
+        @Override
+        public Origin getOrigin() {
+            return nettyConnection.getOrigin();
+        }
+
+        @Override
+        public long getTimeToFirstByteMillis() {
+            return nettyConnection.getTimeToFirstByteMillis();
+        }
+
+        @Override
+        public void addConnectionListener(Listener listener) {
+            nettyConnection.addConnectionListener(listener);
+        }
+
+        @Override
+        public void close() {
+            nettyConnection.close();
+        }
+
+        @Override
+        public boolean isExpired() {
+            return stopwatch.elapsed(SECONDS) >= connectionExpirationSeconds;
+        }
+    }
```

Comment:
I think this class could perform the expiration check in the `isConnected()` method. Like so:

```
    if (isExpired()) {
        close();
        return true;
    } 
   ...
```

Then we could get rid of the above `shouldTerminate` method, and perhaps we could get rid of `ExpirableConnection` interface too?